tests/unittests/graphOptzTest.cpp

```diff
@@ -1114,3 +1114,39 @@ TEST_F(GraphOptz, concatReshapes) {
       llvm::dyn_cast<ConcatNode>(O->getInput()->getNthInput(0)->getNthInput(0));
   ASSERT_TRUE(newCN);
 }
+
+/// Check that Variable CSE works correctly, combining small Variables that have
+/// the same data.
+TEST_F(GraphOptz, VarsCSE) {
+  // Create three variables that are Private, have TrainKind None, and have no
+  // writers. The first two variables have the same data, and so should be
+  // combined via variable CSE. The third variable differs by the last value,
+  // and so should not be combined.
+  auto *input1 =
+      mod_.createVariable(ElemKind::FloatTy, {10}, "input1",
+                          VisibilityKind::Private, Variable::TrainKind::None);
+  auto *input2 =
+      mod_.createVariable(ElemKind::FloatTy, {10}, "input2",
+                          VisibilityKind::Private, Variable::TrainKind::None);
+  auto *input3 =
+      mod_.createVariable(ElemKind::FloatTy, {10}, "input3",
+                          VisibilityKind::Private, Variable::TrainKind::None);
+  input1->getHandle() = {0, 1, 2, 3, 4, 5, 6, 7, 8, 9};
+  input2->getHandle() = {0, 1, 2, 3, 4, 5, 6, 7, 8, 9};
+  input3->getHandle() = {0, 1, 2, 3, 4, 5, 6, 7, 8, -1};
+
+  // Input them each to different nodes, so node CSE does not change them.
+  auto *TN = F_->createTanh("tanh", input1);
+  auto *SN = F_->createSigmoid("sigmoid", input2);
+  auto *RN = F_->createRELU("relu", input3);
+  auto *CN = F_->createConcat("concat", {TN, SN, RN}, /* axis */ 0);
+  F_->createSave("ret", CN);
+
+  // Initially there are inputs 1, 2, and 3, and the variable for the save.
+  EXPECT_EQ(mod_.getVars().size(), 4);
+
+  ::glow::optimize(F_, CompilationMode::Infer);
+
+  // Now input1 and input2 have been combined, but input 3 has not.
+  EXPECT_EQ(mod_.getVars().size(), 3);
```

Comment:
Could you check that we are indeed left with the variables we expect?
Also, check that the users have been properly updated.

Reply:
Done.